idx/memory/priority_lock.go

```diff
@@ -0,0 +1,124 @@
+package memory
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// PriorityRWMutex is a mutex that allows fair access to high priority read operations while
+// blocking low priority operations if a write lock is waiting.
+type PriorityRWMutex struct {
+	lock        sync.RWMutex
+	lowPrioLock sync.RWMutex
+}
+
+type BlockContext struct {
```

Comment:
can you document this? it's quite hard to reverse engineer this code.
why do callers call Lock on the prioritymutex but then unlock on the blockcontext? why does the index use PriorityRWMutex but then the BlockContext seems to embed this same PriorityRWMutex ?

Reply:
> can you document this?

Will do.

> why do callers call Lock on the prioritymutex but then unlock on the blockcontext?

The intent of going through the BlockContext was to not put the logging calls into the PriorityRWMutex itself and also deduplicate some of the code from the site of the unlock call. Technically, we could wrap both calls into the BlockContext so that callers do something like:

```
bc := NewBlockContext(lock)
bc.RLockLow()
...critical section code...
bc.UnlockLow(...)
```

But now we need two lines so we would probably want to do a couple of functions like `bc := LockLow(readLock)` and now we are basically back to what we have. 
Note that the lock is embedded, so extracting the lock type is a little ugly.

We *could* just have the caller pass the BlockContext back to the `Unlock` or `RUnlockLow` call, but it makes it a little more likely (IMO) that the caller might do it incorrectly  by passing in a dummy BlockContext. The goal was to try make the callsite not  look too bad.

> why does the index use PriorityRWMutex but then the BlockContext seems to embed this same PriorityRWMutex ?

How else would the BlockContext now which PriorityRWMutex to unlock? `BlockContext` is a *short-term* variable. Designed to only be held for the duration of lock acquisition.

Reply:
ok i understand this now. it makes sense. The general layout seems fine.
My main gripe is , it seems we pretty much always want to discourage callers calling (please confirm):
* `PriorityRWMutex.RUnlock` & `PriorityRWMutex.RUnlockLow`. Instead, should call `BlockContext.RUnlockLow`
* `PriorityRWMutex.Unlock` // instead, should call `BlockContext.Unlock`

if we could take these 3 methods and lowercase their first letter, that would clarify it more that it's an internal method.
For the RUnlock/RUnlockLow, that works fine, but with Unlock, there's a problem. Because we have 2 special cases in the code like so:

```
		// it is important to release the lock for a short time, otherwise
		// it's possible that the enricher can't process its queue because
		// it could be blocked on waiting for the read lock on the index
		// which would lead to deadlock.
		m.Unlock(&BlockContext{})
```

I see how it makes sense, it's just one of these things that makes this code more confusing.

Reply:
>  it seems we pretty much always want to discourage callers calling (please confirm)

Confirmed!

> with Unlock, there's a problem. Because we have 2 special cases in the code like so

Yes, and that interaction is actually point of pain (see #1881 for reducing the likelihood of that happening). Ideally there would be no code that needs to do that (IMO, that shows that the interaction of these components is too complex and needs to be refactored). However, your point stands. 

I can unexport those functions (even `Unlock`) and things will work within the memory package (provided this lock is never factored out).

Reply:
right. we may factor this out of the package at some point..
perhaps the best solution is simply to document for each function whether they should be called directly (or under which circumstance) and when to call the BlockContext version instead.